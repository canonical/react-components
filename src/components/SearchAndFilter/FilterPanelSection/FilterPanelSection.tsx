import React, { useEffect, useRef, useState } from "react";
import Chip from "../../Chip";
import { overflowingChipsCount, isChipInArray } from "../utils";
import { highlightSubString } from "../../../utils";
import type { SearchAndFilterChip, SearchAndFilterData } from "../types";

export type Props = {
  /**
   * The chip data to display in the panel.
   */
  data: SearchAndFilterData;
  /**
   * The current search chips.
   */
  searchData: SearchAndFilterChip[];
  /**
   * A search string.
   */
  searchTerm: string;
  /**
   * Whether the chips should be hidden.
   */
  sectionHidden?: boolean;
  /**
   * A function to toggle whether a chip is selected.
   */
  toggleSelected: (chip: SearchAndFilterChip) => void;
};

const FilterPanelSection = ({
  data,
  searchData,
  searchTerm = "",
  sectionHidden,
  toggleSelected,
}: Props): JSX.Element => {
  const { chips, heading } = data;
  const [overflowCounter, setOverflowCounter] = useState(0);
  const [expanded, setExpanded] = useState(false);
  const chipWrapper = useRef(null);

  const handleChipClick = (chip) => {
    toggleSelected(chip);
  };

  // If the offsetTop is more than double height of a single chip, consider it
  // overflowing
  const updateFlowCount = function () {
    const chips = chipWrapper?.current?.querySelectorAll(".p-chip");
    const overflowCount = overflowingChipsCount(chips, 2);
    setOverflowCounter(overflowCount);
  };

  // Check if search term characters matches any characters in panel heading
  const searchTermInHeading = highlightSubString(heading, searchTerm).match;

  // Serialise chip values into string so it can be interrogated with subString
  const chipValues = [];
  Object.entries(chips).forEach((chipValue) => {
    chipValues.push(chipValue[1].value);
  });

  // Search chips for character match with search term
  const searchTermInChips = highlightSubString(
    chipValues.toString(),
    searchTerm
  ).match;

  const panelSectionVisible =
    searchTermInHeading || searchTermInChips || searchTerm === "";

  // Update overflow count when component is resized
  useEffect(() => {
    const resizeObserverSupported = typeof ResizeObserver !== "undefined";
    const wrapper = chipWrapper?.current;
    let wrapperWidthObserver: ResizeObserver;
    if (resizeObserverSupported && panelSectionVisible) {
      wrapperWidthObserver = new ResizeObserver(() => {
        updateFlowCount();
      });
      wrapperWidthObserver.observe(wrapper);
    } else {
      updateFlowCount();
    }
    return () => {
      resizeObserverSupported && wrapperWidthObserver?.disconnect();
    };
  }, [panelSectionVisible]);

  // When overflow counter is clicked, all chips are shown
  const showAllChips = () => {
    setExpanded(true);
  };

  return (
    <>
      {panelSectionVisible && (
        <div className="p-filter-panel-section">
          {heading && chips.length > 0 && (
            <h3
              className="p-filter-panel-section__heading"
              dangerouslySetInnerHTML={{
                __html: highlightSubString(heading, searchTerm).text,
              }}
            />
          )}
          <div
            className="p-filter-panel-section__chips"
            aria-expanded={expanded}
            ref={chipWrapper}
          >
            {chips?.map((chip) => {
              // If search term has been added to input, only matching chips
              // should display
              const searchTermInChip = highlightSubString(
                chip.value,
                searchTerm
              ).match;
              const chipVisible =
                searchTermInChip ||
                searchTerm === "" ||
                highlightSubString(heading, searchTerm).match;
              return (
                chipVisible &&
                !sectionHidden && (
                  <Chip
                    key={`${chip.lead}+${chip.value}`}
                    lead={chip.lead}
                    value={chip.value}
                    selected={isChipInArray(chip, searchData)}
                    subString={searchTerm}
                    onClick={() => handleChipClick(chip)}
                  />
                )
              );
            })}
            {overflowCounter > 0 && !expanded && (
              <span
                className="p-filter-panel-section__counter"
                onClick={showAllChips}
                onKeyPress={showAllChips}
                tabIndex={0}
              >
                +{overflowCounter}
              </span>
            )}
          </div>
        </div>
      )}
    </>
  );
};

export default FilterPanelSection;
