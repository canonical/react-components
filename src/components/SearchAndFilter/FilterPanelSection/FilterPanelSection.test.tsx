import { render, screen } from "@testing-library/react";
import React from "react";

import FilterPanelSection from "./FilterPanelSection";
import userEvent from "@testing-library/user-event";

const sampleData = {
  id: 1,
  heading: "Regions",
  chips: [{ value: "us-east1" }, { value: "us-east2" }, { value: "us-east3" }],
};

describe("Filter panel section", () => {
  it("renders", () => {
    render(
      <FilterPanelSection
        searchData={[]}
        searchTerm=""
        toggleSelected={jest.fn()}
        data={sampleData}
      />
    );
    // eslint-disable-next-line testing-library/no-node-access
    expect(document.querySelector(".p-filter-panel-section")).toMatchSnapshot();
  });

  it("shows correct data passed as prop", () => {
    render(
      <FilterPanelSection
        searchData={[]}
        searchTerm=""
        toggleSelected={jest.fn()}
        data={sampleData}
      />
    );
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-filter-panel-section")
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "us-east1" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "us-east2" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("button", { name: "us-east3" })
    ).toBeInTheDocument();
    expect(
      screen.getByRole("heading", { name: "Regions" })
    ).toBeInTheDocument();
  });

  it("should hide chip overflow counter when none overflow", () => {
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 40,
    });
    render(
      <FilterPanelSection
        searchData={[]}
        searchTerm=""
        toggleSelected={jest.fn()}
        data={sampleData}
      />
    );
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-search-and-filter__selected-count")
    ).not.toBeInTheDocument();
  });

  it("show overflow chip counter when chips overflow", () => {
    // Jest is unaware of layout so we must mock the offsetTop and offsetHeight
    // of the chips
    Object.defineProperty(HTMLElement.prototype, "offsetHeight", {
      configurable: true,
      value: 40,
    });
    Object.defineProperty(HTMLElement.prototype, "offsetTop", {
      configurable: true,
      value: 100,
    });
    render(
      <FilterPanelSection
        searchData={[]}
        searchTerm=""
        toggleSelected={jest.fn()}
        data={sampleData}
      />
    );
    expect(
      // Use a query selector because the element's text is split up over
      // multiple elements so it can't be selected by its content.
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-filter-panel-section__counter")?.textContent
    ).toBe("+3");
  });

  it("all chips are shown when counter is clicked", async () => {
    render(
      <FilterPanelSection
        searchData={[]}
        searchTerm=""
        toggleSelected={jest.fn()}
        data={sampleData}
        sectionHidden={false}
      />
    );
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-filter-panel-section__chips")
    ).toHaveAttribute("aria-expanded", "false");
    await userEvent.click(
      // Use a query selector because the element's text is split up over
      // multiple elements so it can't be selected by its content.
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-filter-panel-section__counter") as HTMLElement
    );
    expect(
      // eslint-disable-next-line testing-library/no-node-access
      document.querySelector(".p-filter-panel-section__chips")
    ).toHaveAttribute("aria-expanded", "true");
  });
});
