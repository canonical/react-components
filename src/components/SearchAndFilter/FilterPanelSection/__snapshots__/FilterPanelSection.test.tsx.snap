// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Filter panel section renders 1`] = `
<Fragment>
  <div
    className="p-filter-panel-section"
  >
    <h3
      className="p-filter-panel-section__heading"
      dangerouslySetInnerHTML={
        Object {
          "__html": "Regions",
        }
      }
    />
    <div
      aria-expanded={false}
      className="p-filter-panel-section__chips"
    >
      <Chip
        key="undefined+us-east1"
        onClick={[Function]}
        selected={false}
        subString=""
        value="us-east1"
      />
      <Chip
        key="undefined+us-east2"
        onClick={[Function]}
        selected={false}
        subString=""
        value="us-east2"
      />
      <Chip
        key="undefined+us-east3"
        onClick={[Function]}
        selected={false}
        subString=""
        value="us-east3"
      />
    </div>
  </div>
</Fragment>
`;
